cmd/influxd/restore.go

```diff
@@ -0,0 +1,257 @@
+	// Create root directory.
+	if err := os.MkdirAll(path, 0777); err != nil {
+		return fmt.Errorf("mkdir: err=%s", err)
+	}
+
+	// Loop over files and extract.
+	for {
+		// Read entry header.
+		sf, err := sr.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return fmt.Errorf("next: entry=%s, err=%s", sf.Name, err)
+		}
+
+		// Create parent directory for output file.
+		if err := os.MkdirAll(filepath.Dir(filepath.Join(path, sf.Name)), 0777); err != nil {
+			return fmt.Errorf("mkdir: entry=%s, err=%s", sf.Name, err)
+		}
+
+		if err := func() error {
+			// Create output file.
+			f, err := os.Create(filepath.Join(path, sf.Name))
+			if err != nil {
+				return fmt.Errorf("create: entry=%s, err=%s", sf.Name, err)
+			}
+			defer f.Close()
+
+			// Copy contents from reader.
+			if _, err := io.CopyN(f, sr, sf.Size); err != nil {
+				return fmt.Errorf("copy: entry=%s, err=%s", sf.Name, err)
+			}
+
+			return nil
+		}(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// materialize creates broker & raft directories based on the snapshot.
+func (cmd *RestoreCommand) materialize(path string, ss *influxdb.Snapshot, u url.URL) error {
+	// Materialize broker.
+	if err := cmd.materializeBroker(path, ss.Index()); err != nil {
+		return fmt.Errorf("broker: %s", err)
+	}
+
+	// Materialize raft.
+	if err := cmd.materializeRaft(filepath.Join(path, "raft"), u); err != nil {
+		return fmt.Errorf("raft: %s", err)
+	}
+
+	return nil
+}
+
+func (cmd *RestoreCommand) materializeBroker(path string, index uint64) error {
+	// Create root directory.
+	if err := os.MkdirAll(path, 0777); err != nil {
+		return fmt.Errorf("mkdir: err=%s", err)
+	}
+
+	// Create broker meta store.
+	meta, err := bolt.Open(filepath.Join(path, "meta"), 0666, &bolt.Options{Timeout: 1 * time.Second})
+	if err != nil {
+		return fmt.Errorf("open broker meta: %s", err)
+	}
+	defer meta.Close()
+
+	// Write highest index to meta store.
+	if err := meta.Update(func(tx *bolt.Tx) error {
+		b, err := tx.CreateBucket([]byte("meta"))
+		if err != nil {
+			return fmt.Errorf("create meta bucket: %s", err)
+		}
+
+		if err := b.Put([]byte("index"), u64tob(index)); err != nil {
+			return fmt.Errorf("put: %s", err)
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("update broker meta: %s", err)
+	}
+
+	return nil
+}
+
+func (cmd *RestoreCommand) materializeRaft(path string, u url.URL) error {
+	// Create raft directory.
+	if err := os.MkdirAll(path, 0777); err != nil {
+		return fmt.Errorf("mkdir raft: err=%s", err)
+	}
+
+	// Write raft id & term.
+	if err := ioutil.WriteFile(filepath.Join(path, "id"), []byte(`1`), 0666); err != nil {
+		return fmt.Errorf("write raft/id: %s", err)
+	}
+	if err := ioutil.WriteFile(filepath.Join(path, "term"), []byte(`1`), 0666); err != nil {
+		return fmt.Errorf("write raft/term: %s", err)
+	}
+
+	// Generate configuration.
+	var rc raft.Config
+	rc.ClusterID = uint64(rand.Int())
```

Comment:
Yes, that's correct. We don't snapshot the broker so we don't have the cluster id but it's also good so that we don't accidentally restore to a new cluster and have those two clusters try to stomp on each other. The backup/restore is for a catastrophic cluster failure.